driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/AggregatePublisherImpl.java

```diff
@@ -16,87 +16,195 @@
 
 package com.mongodb.reactivestreams.client.internal;
 
+import com.mongodb.MongoNamespace;
+import com.mongodb.ReadConcern;
+import com.mongodb.ReadPreference;
+import com.mongodb.WriteConcern;
 import com.mongodb.client.model.Collation;
-import com.mongodb.internal.async.client.AsyncAggregateIterable;
+import com.mongodb.internal.async.AsyncBatchCursor;
+import com.mongodb.internal.async.client.OperationExecutor;
+import com.mongodb.internal.async.client.WriteOperationThenCursorReadOperation;
+import com.mongodb.internal.client.model.AggregationLevel;
+import com.mongodb.internal.client.model.FindOptions;
+import com.mongodb.internal.operation.AsyncOperations;
+import com.mongodb.internal.operation.AsyncReadOperation;
+import com.mongodb.internal.operation.AsyncWriteOperation;
+import com.mongodb.lang.Nullable;
 import com.mongodb.reactivestreams.client.AggregatePublisher;
+import com.mongodb.reactivestreams.client.ClientSession;
+import org.bson.BsonDocument;
+import org.bson.BsonString;
+import org.bson.codecs.configuration.CodecRegistry;
 import org.bson.conversions.Bson;
 import org.reactivestreams.Publisher;
-import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Mono;
 
+import java.util.List;
 import java.util.concurrent.TimeUnit;
 
 import static com.mongodb.assertions.Assertions.notNull;
-
-
-final class AggregatePublisherImpl<TResult> implements AggregatePublisher<TResult> {
-
-    private final AsyncAggregateIterable<TResult> wrapped;
-
-    AggregatePublisherImpl(final AsyncAggregateIterable<TResult> wrapped) {
-        this.wrapped = notNull("wrapped", wrapped);
+import static com.mongodb.reactivestreams.client.internal.PublisherCreator.sinkToCallback;
+
+final class AggregatePublisherImpl<D, T> extends BatchCursorPublisherImpl<T> implements AggregatePublisher<T> {
+    private final AsyncOperations<D> operations;
+    private final MongoNamespace namespace;
+    private final Class<D> documentClass;
+    private final Class<T> resultClass;
+    private final CodecRegistry codecRegistry;
+    private final List<? extends Bson> pipeline;
+    private final AggregationLevel aggregationLevel;
+    private Boolean allowDiskUse;
+    private long maxTimeMS;
+    private long maxAwaitTimeMS;
+    private Boolean bypassDocumentValidation;
+    private Collation collation;
+    private String comment;
+    private Bson hint;
+
+    AggregatePublisherImpl(@Nullable final ClientSession clientSession, final MongoNamespace namespace,
```

Comment:
All publishers no longer wrap, so the logic has been lifted to high (the reactive streams driver).